tsdb/engine/tsm1/compact.go

```diff
@@ -406,17 +407,49 @@ func (c *DefaultPlanner) findGenerations() tsmGenerations {
 // Compactor merges multiple TSM files into new files or
 // writes a Cache into 1 or more TSM files
 type Compactor struct {
-	Dir    string
-	Cancel chan struct{}
-	Size   int
+	Dir  string
+	Size int
 
 	FileStore interface {
 		NextGeneration() int
 	}
+
+	mu      sync.RWMutex
+	opened  bool
+	closing chan struct{}
+}
+
+func (c *Compactor) Open() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.opened {
+		return
+	}
+
+	c.closing = make(chan struct{})
+	c.opened = true
+}
+
+func (c *Compactor) Close() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if !c.opened {
```

Comment:
I tried that initially, but I need to be able to re-open the compactor which means I'm setting `c.closing` again. That requires taking a lock so `select` + `lock` seemed more complicated than just taking the lock.